### ISSUE TITLE
Visual Studio Code & Source Maps

### ISSUE THREAD

**Issue by tomitrescak** [meteor/meteor#7335]
Hi, I was trying to setup the debug session with Visual Studio Code and Chrome Debug Tools. The whole process fails, as Meteor is using proprietary way of serving source maps via `meteor://` paths (reported by @roblourens from MS in https://github.com/Microsoft/vscode-chrome-debug/issues/183#issuecomment-229818373).

Is there any way how can I change the source maps mappings either to physical files n HD or just setup different `sources` or `sourcesRoot` in order to allow Chrome (even remote) debugging?

Thanks

**Comment by abernix** [comment]
I'm not sure how Visual Studio Code comes into play here (though, I'm admittedly unfamiliar with its offerings).  However, I'm able to use source maps on a fresh project using Chrome without any problems – and no additional configuration.  You should be able to debug in Chrome without Visual Studio Code at all.

Regardless, you haven't provided enough information here to begin to help though.  What version of Meteor are you experiencing this problem on?  Does this problem occur to you on a fresh Meteor project?  The source maps are exactly what _does_ provide the mapping between `meteor://` file paths and physical files – as you can see here on one of my projects where "💻 app" is actually representing `meteor://app/` paths.  Without source map, this wouldn't be showing:

<img width="380" alt="screen shot 2016-07-04 at 3 37 09 pm" src="https://cloud.githubusercontent.com/assets/841294/16560613/36f89cee-41fd-11e6-83df-89cfe874d986.png">

Does debugging directly in Chrome not work for you?  Do you not see this in your _Sources_ tab of Chrome?

**Comment by tomitrescak** [comment]
@abernix I'm using this with the latest version of Meteor and version 1.3.4.1.
I can reproduce with the fresh project.

My issue is that I have quite large Meteor projects and due to the way Meteor is packing everything into couple large files (CSS and JS), Chrome Debugging becomes really slow. After page refresh hitting breakpoint takes up to 5-10 seconds to show me the source code, meanwhile displaying "Waiting for source maps message". I have one of the latest Macbooks pro with 16GB of RAM and SSD drive, so computer performance is not an issue.

I found out that it tis possible to debug code in VS Code using Chrome Debug Tools, so I wanted to give it a go. I confirm that ChromeDebugging works, but it becomes quite slow for larger projects.

**Comment by abernix** [comment]
Whether using Virtual Studio Code or not, when you make a change to a single file, that file is still going to be compiled into a single, combined `.js`.  A source map is written in parallel and hinted to the debugger (using a HTTP header). 

Do you have any reason or explanation to believe that it would be faster from within Virtual Studio Code?  I'm imagining that it is simply using the same source map delivered by Meteor to Chrome (but for some reason VSC doesn't know how to interpret what Meteor is sending).

**Comment by tomitrescak** [comment]
@abernix I am just not happy with the current state so I am trying anything possible :) Since this is a thick client, I believe that there can be some performance gains ... maybe not. Microsoft team reported that the problem is the URL style for Meteor apps, so I was wondering if this can be changed.

**Comment by abernix** [comment]
I understand. #7076 is a general issue open to hopefully improve source mapping.  Unfortunately, the sourcemap spec is not super mature.  In fact, the official spec that Google and Mozilla refer to is [just a Google Doc](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit).  (Granted, it hasn't changed in a couple years.)

Best I can interpret though, Meteor is not incorrectly generating the source map.  The [official Mozilla Source Map npm module](https://www.npmjs.com/package/source-map) consumes the map Meteor generates without problem.

https://github.com/Microsoft/vscode-chrome-debug/issues/183#issuecomment-229818373 hinted that they thought Meteor was in the wrong, but their correct solution was indicated to be a "workaround for the meantime" – suggesting that they need to fix something as well.

Anyhow, it's all a bit unclear so I left a request for input here if you want to take a read: https://github.com/Microsoft/vscode-chrome-debug/issues/183#issuecomment-230454757

**Comment by streamnsight** [comment]
I can confirm that Chrome debug is PAINFULY slow with a large project, which lead me to this issue... 

Is there a way to speed things up here? it takes like 5-10sec to load the code, but also make Chrome almost unusable meanwhile (I can't even seem to resize the various windows in less then 2-3 sec at any time.)

From other reading on this type of slowness, i hear it may be due to having large projects in Chrome 'workspace'
I don't use workspaces, but I am wondering if Meteor somewhat uses this implicitly, as I can find my code individual files/folder structure under >app ...

any input on how to troubleshoot this kind of thing would be greatly appreciated!

**Comment by abernix** [comment]
@streamnsight Meteor does not force Workspaces.  Sourcemaps are, as I said above, not super refined and while there have been improvements, there are still issues that haven't been ironed out (outside of Meteor).  Is your development server running locally or are you accessing it over the network?  Sourcemaps can, in general, be more than twice the size of your code itself so this can cause problems as your browser has to refetch them when they change.  They can also take time to build on larger files, so it's best to have many smaller files instead of single large files as it will allow only small parts to be rebuilt.

**Comment by streamnsight** [comment]
@abernix 
to clarify: meteor is local, project is a lot of small files, not large files.
chrome devtools window gets really slow and uses up to 100% CPU sometimes. If not related to workspaces, something is slowing down chrome devtool tab, and it has to do with Meteor's ways of handling things as i also use it for even larger Angular projects and i don't have this problem.

**Comment by abernix** [comment]
@streamnsight Are the larger Angular projects using Source Maps?  Do you notice speed improvements if you disable Source Maps in Chrome DevTools?  You can do this by going to:

  Preferences _(shortcut: F1)_ **>** Sources **>** Uncheck _"Enable JavaScript source maps"_

**Comment by abernix** [comment]
I'm going to close this due to inactivity and also because per https://github.com/Microsoft/vscode-chrome-debug/issues/183#issuecomment-231487834 there were changes made to Visual Studio Code (as of 0.4.7) so it would work properly with Meteor (and other frameworks).  Any other issues with Chrome would be a separate issue and there are other issues open for that (granted other source map problems have been fixed recently).

**Comment by tomitrescak** [comment]
@abernix even with all the latest improvements it was not working with Meteor. I did not work on meteor project in past months so if anything changed, good.
